<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="arm">
    <link name = "axis">
    <visual>
        <origin xyz="0 0 0" rpy="0 0 0" />
    </visual>
    </link>

    <xacro:property name="base_len" value="0.5" />
    <link name="base_link">
    <visual>
        <origin xyz="0 0 ${base_len/2.0}" rpy="0 0 0" />
        <geometry>
        <cylinder radius="0.2" length="${base_len}" />
        </geometry>
    </visual>
    </link>

  <joint name="axis_to_base" type="fixed">
    <parent link="axis"/>
    <child link="base_link"/>
  </joint>

  <xacro:property name="l1_len" value="1.0" />
  <link name="link1">
    <visual>
        <origin xyz="0 0 ${l1_len/2.0}" rpy="0 0 0" />
        <geometry>
        <cylinder radius="0.1" length="${l1_len}" />
        </geometry>
       
    </visual>
  </link>


   <joint name="base_to_link1" type="revolute">
   <axis xyz="0 0 1"/>
   <origin xyz="0 0 ${base_len}" rpy="0 0 0"/>
   <limit effort="10" velocity="1" lower="-3.14" upper="3.14"/>
    <parent link="base_link"/>
    <child link="link1"/>
  </joint>

  <xacro:property name="l2_len" value="1.0" />
  <link name="link2">
    <visual>
        <origin xyz="${l2_len/2.0} 0 0" rpy="0 1.57 0" />
        <geometry>
        <cylinder radius="0.1" length="${l2_len}" />
        </geometry>
       
    </visual>
  </link>

   <joint name="link1_to_link2" type="revolute">
   <axis xyz="0 1 0"/>
   <origin xyz="0 0 ${l1_len}" rpy="0 0 0"/>
   <limit effort="10" velocity="1" lower="-3.14" upper="3.14"/>
    <parent link="link1"/>
    <child link="link2"/>
  </joint>

  <xacro:property name="l3_len" value="0.3" />
  <link name="link3">
    <visual>
        <origin xyz="${l3_len/2.0} 0 0" rpy="0 1.57 0" />
        <geometry>
        <cylinder radius="0.1" length="${l3_len}" />
        </geometry>
       
    </visual>
  </link>

   <joint name="link2_to_link3" type="revolute">
   <axis xyz="0 1 0"/>
   <origin xyz="${l2_len} 0 0" rpy="0 0 0"/>
   <limit effort="10" velocity="1" lower="-3.14" upper="3.14"/>
    <parent link="link2"/>
    <child link="link3"/>
  </joint>

  <xacro:property name="l4_len" value="0.3" />
  <xacro:property name="l4_width" value="0.2" />
  <link name="link4">
    <visual>
        <origin xyz="${l4_len/2.0} 0 0" rpy="1.57 0 0" />
        <geometry>
        <cylinder radius="${l4_len/2}" length="${l4_width}" />
        </geometry>
    </visual>
  </link>

   <joint name="link3_to_link4" type="revolute">
   <axis xyz="1 0 0"/>
   <origin xyz="${l3_len} 0 0" rpy="0 0 0"/>
   <limit effort="10" velocity="1" lower="-3.14" upper="3.14"/>
    <parent link="link3"/>
    <child link="link4"/>
  </joint>

  <xacro:property name="l5_len" value="0.2" />
  <link name="link5">
    <visual>
        <origin xyz="0 0 -${l5_len/2.0}" rpy="0 0 0" />
        <geometry>
        <cylinder radius="0.1" length="${l5_len}" />
        </geometry>
    </visual>
  </link>

  <joint name="link4_to_link5_pitch" type="revolute">
   <axis xyz="0 1 0"/>
   <origin xyz="${l4_len} 0 0" rpy="0 0 0"/>
   <limit effort="10" velocity="1" lower="-3.14" upper="3.14"/>
    <parent link="link4"/>
    <child link="link5"/>
  </joint>

  <link name="link5_roll">
  </link>

  <joint name="link4_to_link5_roll" type="revolute">
   <axis xyz="0 0 1"/>
   <origin xyz="0 0 0" rpy="0 0 0"/>
   <limit effort="10" velocity="1" lower="-3.14" upper="3.14"/>
    <parent link="link5"/>
    <child link="link5_roll"/>
  </joint>

  <link name="end_effector">
  </link>

  <joint name="link5_to_end" type="fixed">
   <origin xyz="0 0 -${l5_len}" rpy="0 0 0"/>
    <parent link="link5_roll"/>
    <child link="end_effector"/>
  </joint>

</robot>